terraform/ec2-examples/distributed-ml-training-fsdp/README.md

```diff
@@ -0,0 +1,143 @@
+```
+
+2. Connect to the container
+
+Due to the size of the container images, it might take several minutes until the containers reach a running state. The following command will fail if the container is not running.
+
+```
+CONTAINER_ID=$(sudo docker ps -qf "name=.*-rayhead-.*")
+sudo docker exec -it $CONTAINER_ID bash
+```
+
+3. Inside the container shell, check the cluster status. 3 nodes should be listed as healthy with 2.0 GPUs available - If you do not see 2.0 GPUs, the workers have not started yet.
+
+```bash
+ray status
+```
+
+Example output:
+
+```
+(…)
+Node status
+---------------------------------------------------------------
+Active:
+ 1 node_ae3cba0ce4b9196f86f23efaa3b5cf53cb3956d96162d0965e901cf5
+ 1 node_0a46d6a2d9b0f1f88f0b3ba7bc0237e512e5435c9cab454e20c95f52
+ 1 node_7839a944b0cc9213137490d8cccb7bd7eb4c9fbc8b6f56123532bdc7
+(…)
+
+```
+
+4. Run the [training script example](./training_example.py) - It uses fully sharded data parallel to split the data between GPUs. You can look at the comments inside the python script to learn more about each step. A bucket is created as part of the terraform plan (Bucket name is printed as output). Make sure to add the name of that bucket (starts with "dt-results-") as argument of the training_example.py script
+
+```bash
+cd /tmp
+wget https://raw.githubusercontent.com/aws-ia/ecs-blueprints/main/terraform/ec2-examples/distributed-ml-training-fsdp/training_example.py
+python training_example.py REPLACE_WITH_YOUR_BUCKET_NAME
+```
+
+Example output:
+
+```
+(RayTrainWorker pid=364, ip=10.0.0.218) Setting up process group for: env:// [rank=0, world_size=12]
+(TorchTrainer pid=287, ip=10.0.0.218) Started distributed worker processes:
+(TorchTrainer pid=287, ip=10.0.0.218) - (node_id=47a3d7c20ebb914a34340d9a4cdd08d8b8f35051eb7d1821431e0d97, ip=10.0.0.218, pid=364) world_rank=0, local_rank=0, node_rank=0
+(…)
+Epoch 0: : 1it [00:26, 26.64s/it, v_num=0, train_loss=13.00]
+(…)
+Epoch 0: : 129it [43:14, 20.11s/it, v_num=0, train_loss=0.162]
+(…)
+╭─────────────────────────────────────────╮
+│ Training result                         │
+├─────────────────────────────────────────┤
+│ checkpoint_dir_name   checkpoint_000000 │
+│ time_this_iter_s             2968.97071 │
+│ time_total_s                 2968.97071 │
+│ training_iteration                    1 │
+│ epoch                                 0 │
+│ step                                129 │
+│ train_loss                      0.16882 │
+╰─────────────────────────────────────────╯
+(…)
+Training finished iteration 1 at 2025-01-10 19:56:58. Total running time: 51min 37s
+
+```
+
+It can take approximately an hour for the scrip to finish. The terraform plan deploys a custom dashboard in CloudWatch  named "distributed-ml-training-fsdp" with memory and utilization of the GPUs in the cluster. Once the training finishes, it should show GPUs utilized near 100% and memory between 10 and 20 GB used.
+
+## Clean up
```

Comment:
Before cleanup can you explain what the goal of the script we executed ?

Reply:
Yes. Added an additional comment on what the script does.

Reply:
Thanks, it's more clerer now, can you go a bit further and explain use case for using the generated model ?

Reply:
Thanks. Included more information this time about the resulting model